test/functorch/test_control_flow.py

```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+            torch._dynamo.exc.Unsupported,
+            "Observed exception.*",
+        ):
+            result = scan(fct_wrong_pytree, init=init, input=inp, dim=0)
+
+        result = scan(fct_pointwise, init=init, input=inp, dim=0, reverse=reverse)
+        expected_result = _fake_scan(
+            fct_pointwise, init=init, input=inp, dim=0, reverse=reverse
+        )
+        self.assertEqual(result, expected_result)
+
+    @unittest.skipIf(not SM70OrLater, "triton")
+    @requires_cuda
+    @parametrize("combine_mode", ["pointwise", "generic"])
+    @parametrize("compile_mode", ["none", "eager", "compile", "compile_dynamic_shape"])
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    # Skipping the combination of combine_mode=pointwise and device=cpu
+    # as the current implementation of pointwise does only support CUDA device
+    @decorateIf(
+        unittest.skip,
+        lambda params: (
+            (
+                params["combine_mode"] == "pointwise"
+                and params["device"] == torch.device("cpu")
+            )
+            or params["compile_mode"] == "eager"
+        ),
+    )
+    def test_associative_scan_downstream_op(
+        self, combine_mode, compile_mode, reverse, device
+    ):
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        # Chain with matmul
+        def chain_fct(inp):
+            W = torch.ones(2, 5, device=device)
+            o = associative_scan(
+                add, inp, 1, reverse=reverse, combine_mode=combine_mode
+            )
+            return o @ W
+
         torch.compiler.reset()
-        associative_scan1 = torch.compile(associative_scan, fullgraph=True)
-        associative_scan2 = associative_scan
+        if compile_mode == "compile":
+            fct_cmp = torch.compile(chain_fct, fullgraph=True, dynamic=False)
+        elif compile_mode == "compile_dynamic_shape":
+            fct_cmp = torch.compile(chain_fct, fullgraph=True, dynamic=True)
+        elif compile_mode == "eager":
+            fct_cmp = torch.compile(chain_fct, fullgraph=True, backend="eager")
+        else:
+            fct_cmp = chain_fct
 
-        result1 = associative_scan1(
-            fct_pointwise, inp, 0, combine_mode=combine_mode, reverse=reverse
+        inp = torch.randn(3, 10, 2, device=device)
+        expected_result = _fake_associative_scan(
+            add, inp, 1, reverse=reverse
+        ) @ torch.ones(2, 5, device=device)
+        result1 = fct_cmp(inp)
+        self.assertEqual(result1, expected_result)
+
+        # Chain with scan
+        def chain_fct_same_dim(inp):
+            o1 = associative_scan(
+                add, inp, 1, combine_mode=combine_mode, reverse=reverse
+            )
+            o2 = associative_scan(
+                add, o1, 1, combine_mode=combine_mode, reverse=reverse
+            )
+            return o2
+
+        torch.compiler.reset()
```

Comment:
I'm not sure if we need to reset every time.

Reply:
No, we don't. Removed.